dsl_parser/constraints.py

```diff
@@ -0,0 +1,289 @@
+########
+# Copyright (c) 2019 Cloudify Platform Ltd. All rights reserved
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+#    * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    * See the License for the specific language governing permissions and
+#    * limitations under the License.
+
+import re
+
+from utils import string_types
+from dsl_parser import functions, exceptions
+from constants import INPUTS, DEFAULT, CONSTRAINTS as CONSTRAINT_CONST
+
+_NOT_COMPARABLE_ERROR_MSG = "Value is not comparable, the Constraint " \
+                            "argument type  is '{0}' but value type is '{1}'."
+_NO_LENGTH_ERROR_MSG = "Value's length could not be computed. Value " \
+                       "type is '{0}'."
+
+# Constraint types:
+#  scalar (int/float/long), dual scalar (list/tuple), list (list/tuple),
+#  regex (string)
+_SCALAR, _DUAL_SCALAR, _SEQUENCE, _REGEX = range(4)
+
+CONSTRAINTS = {}
+VALIDATION_FUNCTIONS = {}
+
+
+def register_constraint(constraint_data_type, name, cls=None):
+    if cls is None:
+        def partial(_cls):
+            return register_constraint(
+                constraint_data_type=constraint_data_type, cls=_cls, name=name)
+
+        return partial
+    CONSTRAINTS[name] = cls
+    cls.name = name
+    cls.constraint_data_type = constraint_data_type
+    return cls
+
+
+def register_validation_func(constraint_data_type, f=None):
+    if f is None:
+        def partial(_f):
+            return register_validation_func(constraint_data_type, _f)
+        return partial
+    VALIDATION_FUNCTIONS[constraint_data_type] = f
+    return f
+
+
+def _try_predicate_func(p_func, err_msg):
+    """Tries to return the value of the predicate func `p_func`, if it catches
+     a TypeError it raises a ConstraintException instead.
+
+    :param p_func: the actual predicate function of a constraint, the one that
+     checks if a given value complies with the constraint. E.g. value == 2 for
+     an Equal(2) constraint.
+    :param err_msg: an error message to initialize the ConstraintException
+     with, in case TypeError is raised during the `p_func` runtime.
+    :return: whatever `p_func` returns.
+    """
+    try:
+        return p_func()
+    except TypeError:
+        raise exceptions.ConstraintException(err_msg)
+
+
+class Constraint(object):
+    """An abstract constraint operator class, all classes that implement this
+    one must be registered as the other classes below.
+    """
+
+    # The name of the operator as it appears in the YAML file.
+    name = None
+    # The hyper parameter type of this constraint, what type of arguments it
```

Comment:
what do u mean by 'hyper'?

Reply:
I meant a hyperparamter: https://en.wikipedia.org/wiki/Hyperparameter, like in a neural network

Reply:
Guess this doesn't apply here

Reply:
Fixed to match var name

Reply:
Changed to _data_